mirrord/cli/src/port_forward.rs

```diff
@@ -442,12 +435,271 @@ impl PortForwarder {
     }
 }
 
+pub struct ReversePortForwarder {
+    /// details for traffic mirroring or stealing
+    incoming_mode: IncomingMode,
+    /// communicates with the agent (only TCP supported).
+    agent_connection: AgentConnection,
+    /// associates resolved destination ports with local ports.
+    mappings: HashMap<RemotePort, LocalPort>,
+    /// background task (uses IncomingProxy to communicate with layer)
+    background_tasks: BackgroundTasks<MainTaskId, ProxyMessage, IntProxyError>,
+    /// incoming proxy background task tx
+    incoming_proxy: TaskSender<IncomingProxy>,
+
+    /// true if Ping has been sent to agent.
+    waiting_for_pong: bool,
+    ping_pong_timeout: Instant,
+}
+
+impl ReversePortForwarder {
+    pub(crate) async fn new(
+        agent_connection: AgentConnection,
+        parsed_mappings: Vec<PortOnlyMapping>,
+        network_config: IncomingConfig,
+    ) -> Result<Self, PortForwardError> {
+        let mut mappings: HashMap<RemotePort, LocalPort> =
+            HashMap::with_capacity(parsed_mappings.len());
+
+        if parsed_mappings.is_empty() {
+            return Err(PortForwardError::NoMappingsError());
+        }
+
+        for mapping in &parsed_mappings {
+            // check destinations are unique
+            if mappings.contains_key(&mapping.remote) {
+                // two mappings shared a key thus keys were not unique
+                return Err(PortForwardError::ReversePortMapSetupError(mapping.remote));
+            }
+            mappings.insert(mapping.remote, mapping.local);
+        }
```

Comment:
How about making the type of the port mapping command argument a tuple type that holds a `HashMap`, and do this parsing and verification in the `FromStr` implementation for that type?

Reply:
I am not entirely sure this is possible within clap (see [this discussion](https://github.com/clap-rs/clap/discussions/4291)) - I think this check has to be done in ReversePortForwarder

Reply:
Ok so we can use a `Vec`, but the verification that there are no duplicates should happen early (before creating an agent etc.), and the error message should not say "This is a bug". We could accept a single delimited list (`80:80;8080:8080`) and parse it into a hashmap with `FromStr`, but apparently we want to copy the ssh interface, so we will have to live with verification after parsing.